plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceProvider.java

```diff
@@ -212,10 +228,50 @@ public String getType() {
         return supportedRecipeTypes;
     }
 
+
+
+    /**
+     * Creates instance from scratch or by reusing a previously one by using specified {@link MachineSource}
+     * data in {@link MachineConfig}.
+     *
+     * @param machine
+     *         machine description
+     * @param creationLogsOutput
+     *         output for instance creation logs
+     * @return newly created {@link Instance}
+     * @throws UnsupportedRecipeException
+     *         if specified {@code recipe} is not supported
+     * @throws InvalidRecipeException
+     *         if {@code recipe} is invalid
+     * @throws NotFoundException
+     *         if instance described by {@link MachineSource} doesn't exists
+     * @throws MachineException
+     *         if other error occurs
+     */
     @Override
-    public Instance createInstance(Recipe recipe,
-                                   Machine machine,
-                                   LineConsumer creationLogsOutput) throws MachineException, UnsupportedRecipeException {
+    public Instance createInstance(Machine machine, LineConsumer creationLogsOutput)
+            throws UnsupportedRecipeException, InvalidRecipeException, NotFoundException, MachineException {
+
+        // based on machine source, do the right steps
+        MachineConfig machineConfig = machine.getConfig();
+        MachineSource machineSource = machineConfig.getSource();
+        String type = machineSource.getType();
+
+        // get recipe
+        // - it's a dockerfile type:
+        //    - location defined : download this location and get script as recipe
+        //    - content defined  : use this content as recipe script
+        // - it's an image:
+        //    - use either location or content as image ([registry:port]/<repository-image>[:tag][@digest])
```

Comment:
it looks strange to not support same stuff by using location and content as it is the same value for the user.